File: roles/everything-claude-code/files/merge_settings.go
// merge_settings.go — Analyse ~/.claude/settings.json and produce a safe-merge result.
//
// Usage: merge_settings <settings_path> [ecc_clone_path]
//
// Output (JSON on stdout):
//
//	{"status": "new",        "merged": {...}, "details": []}
//	{"status": "ok",         "merged": {...}, "details": []}
//	{"status": "conflict",   "merged": {...}, "details": ["key: 'old' -> 'new'", ...]}
//	{"status": "parse_error","details": ["<error>"]}
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

const (
	marketplaceName = "everything-claude-code"
	pluginKey       = "everything-claude-code@everything-claude-code"
)

type result struct {
	Status  string         `json:"status"`
	Merged  map[string]any `json:"merged,omitempty"`
	Details []string       `json:"details"`
}

func fatal(r result) {
	out, _ := json.Marshal(r)
	fmt.Println(string(out))
	os.Exit(0)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: merge_settings <settings_path> [ecc_clone_path]")
		os.Exit(1)
	}

	settingsPath := os.Args[1]
	clonePath := ""
	if len(os.Args) > 2 {
		clonePath = os.Args[2]
	}

	desired := buildDesired(clonePath)

	data, err := os.ReadFile(settingsPath)
	if os.IsNotExist(err) {
		fatal(result{Status: "new", Merged: desired, Details: []string{}})
	}
	if err != nil {
		fatal(result{Status: "parse_error", Details: []string{err.Error()}})
	}

	var existing map[string]any
	if err := json.Unmarshal(data, &existing); err != nil {
		fatal(result{Status: "parse_error", Details: []string{err.Error()}})
	}

	var conflicts []string

	// model
	if v, ok := existing["model"]; ok && v != desired["model"] {
		conflicts = append(conflicts, fmt.Sprintf("model: '%v' -> '%v'", v, desired["model"]))
	}

	// env keys
	desiredEnv := desired["env"].(map[string]any)
	if existingEnv, ok := existing["env"].(map[string]any); ok {
		for k, dv := range desiredEnv {
			if ev, ok := existingEnv[k]; ok && ev != dv {
				conflicts = append(conflicts, fmt.Sprintf("env.%s: '%v' -> '%v'", k, ev, dv))
			}
		}
	}

	// marketplace path
	if clonePath != "" {
		if existingMkt, ok := existing["extraKnownMarketplaces"].(map[string]any); ok {
			if mkt, ok := existingMkt[marketplaceName].(map[string]any); ok {
				if src, ok := mkt["source"].(map[string]any); ok {
					if oldPath, ok := src["path"].(string); ok && oldPath != clonePath {
						conflicts = append(conflicts,
							fmt.Sprintf("extraKnownMarketplaces.%s.source.path: '%s' -> '%s'",
								marketplaceName, oldPath, clonePath))
					}
				}
			}
		}

		if existingPlugins, ok := existing["enabledPlugins"].(map[string]any); ok {
			if enabled, ok := existingPlugins[pluginKey].(bool); ok && !enabled {
				conflicts = append(conflicts, fmt.Sprintf("enabledPlugins.%s: false -> true", pluginKey))
			}
		}
	}

	merged := merge(existing, desired, clonePath)

	status := "ok"
	if len(conflicts) > 0 {
		status = "conflict"
	}
	fatal(result{Status: status, Merged: merged, Details: conflicts})
}

func buildDesired(clonePath string) map[string]any {
	d := map[string]any{
		"model": "sonnet",
		"env": map[string]any{
			"MAX_THINKING_TOKENS":              "10000",
			"CLAUDE_AUTOCOMPACT_PCT_OVERRIDE":  "50",
			"CLAUDE_CODE_SUBAGENT_MODEL":       "haiku",
		},
	}
	if clonePath != "" {
		d["extraKnownMarketplaces"] = map[string]any{
			marketplaceName: map[string]any{
				"source": map[string]any{
					"source": "directory",
					"path":   clonePath,
				},
			},
		}
		d["enabledPlugins"] = map[string]any{
			pluginKey: true,
		}
	}
	return d
}

func merge(existing, desired map[string]any, clonePath string) map[string]any {
	merged := make(map[string]any, len(existing))
	for k, v := range existing {
		merged[k] = v
	}

	if _, ok := merged["model"]; !ok {
		merged["model"] = desired["model"]
	}

	desiredEnv := desired["env"].(map[string]any)
	mergedEnv := make(map[string]any)
	if existingEnv, ok := existing["env"].(map[string]any); ok {
		for k, v := range existingEnv {
			mergedEnv[k] = v
		}
	}
	for k, v := range desiredEnv {
		mergedEnv[k] = v
	}
	merged["env"] = mergedEnv

	if clonePath != "" {
		mergedMkt := make(map[string]any)
		if existingMkt, ok := existing["extraKnownMarketplaces"].(map[string]any); ok {
			for k, v := range existingMkt {
				mergedMkt[k] = v
			}
		}
		mergedMkt[marketplaceName] = desired["extraKnownMarketplaces"].(map[string]any)[marketplaceName]
		merged["extraKnownMarketplaces"] = mergedMkt

		mergedPlugins := make(map[string]any)
		if existingPlugins, ok := existing["enabledPlugins"].(map[string]any); ok {
			for k, v := range existingPlugins {
				mergedPlugins[k] = v
			}
		}
		mergedPlugins[pluginKey] = true
		merged["enabledPlugins"] = mergedPlugins
	}

	return merged
}

File: roles/everything-claude-code/files/merge_settings.py
#!/usr/bin/env python3
"""
Analyse ~/.claude/settings.json and produce a safe-merge result.

Usage: merge_settings.py <settings_path>

Output (JSON on stdout):
  {"status": "new",        "merged": {...}, "details": []}
  {"status": "ok",         "merged": {...}, "details": []}
  {"status": "conflict",   "merged": {...}, "details": ["key: 'old' -> 'new'", ...]}
  {"status": "parse_error","details": ["<error>"]}
"""
import json
import os
import sys

DESIRED = {
    "model": "sonnet",
    "env": {
        "MAX_THINKING_TOKENS": "10000",
        "CLAUDE_AUTOCOMPACT_PCT_OVERRIDE": "50",
        "CLAUDE_CODE_SUBAGENT_MODEL": "haiku",
    },
}


def main(settings_path):
    if not os.path.exists(settings_path):
        print(json.dumps({"status": "new", "merged": DESIRED, "details": []}))
        return

    with open(settings_path) as f:
        try:
            existing = json.load(f)
        except json.JSONDecodeError as exc:
            print(json.dumps({"status": "parse_error", "details": [str(exc)]}))
            return

    conflicts = []

    if "model" in existing and existing["model"] != DESIRED["model"]:
        conflicts.append("model: '{}' -> '{}'".format(existing["model"], DESIRED["model"]))

    existing_env = existing.get("env", {})
    for key, val in DESIRED["env"].items():
        if key in existing_env and existing_env[key] != val:
            conflicts.append("env.{}: '{}' -> '{}'".format(key, existing_env[key], val))

    merged = dict(existing)
    if "model" not in merged:
        merged["model"] = DESIRED["model"]
    merged_env = dict(merged.get("env", {}))
    for key, val in DESIRED["env"].items():
        merged_env[key] = val
    merged["env"] = merged_env

    status = "conflict" if conflicts else "ok"
    print(json.dumps({"status": status, "merged": merged, "details": conflicts}))


if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit("Usage: merge_settings.py <settings_path>")
    main(sys.argv[1])
